saleor/core/taxes/avatax/__init__.py

```diff
@@ -0,0 +1,388 @@
+from urllib.parse import urljoin
+
+import requests
+from django.conf import settings
+from django.core.cache import cache
+from django.utils.translation import pgettext_lazy
+from requests.auth import HTTPBasicAuth
+
+from saleor.core.utils import get_company_address
+
+from .. import charge_taxes_on_shipping, include_taxes_in_prices
+
+if TYPE_CHECKING:
+    from ....checkout.models import Checkout
+    from ....order.models import Order
+
+META_FIELD = "avatax"
+
+logger = logging.getLogger(__name__)
+
+
+class TransactionType:
+    INVOICE = "SalesInvoice"
+    ORDER = "SalesOrder"
+
+
+class CustomerErrors:
+    DEFAULT_MSG = pgettext_lazy(
+        "We are not able to calculate taxes for your order. Please try later"
+    )
+    ERRORS = ("InvalidPostalCode", "InvalidAddress", "MissingAddress")
+
+    @classmethod
+    def get_error_msg(cls, error: dict) -> str:
+        error_code = error.get("code")
+        if error_code in cls.ERRORS:
+            return error.get("message", cls.DEFAULT_MSG)
+        return cls.DEFAULT_MSG
+
+
+def get_api_url() -> str:
+    """Based on settings return sanbox or production url"""
+    if settings.AVATAX_USE_SANDBOX:
+        return "https://sandbox-rest.avatax.com/api/v2/"
+    return "https://rest.avatax.com/api/v2/"
+
+
+def api_post_request(
+    url: str,
+    data: Dict[str, Any],
+    username: str = settings.AVATAX_USERNAME_OR_ACCOUNT,
+    password: str = settings.AVATAX_PASSWORD_OR_LICENSE,
+) -> Dict[str, Any]:
+    try:
+        auth = HTTPBasicAuth(username, password)
+        response = requests.post(url, auth=auth, data=json.dumps(data), timeout=2)
+        logger.debug("Hit to Avatax to calculate taxes %s", url)
+    except requests.exceptions.RequestException:
+        logger.warning("Fetching taxes failed %s", url)
+        return {}
+    return response.json()
+
+
+def api_get_request(
+    url: str,
+    username: str = settings.AVATAX_USERNAME_OR_ACCOUNT,
+    password: str = settings.AVATAX_PASSWORD_OR_LICENSE,
+):
+    try:
+        auth = HTTPBasicAuth(username, password)
+        response = requests.get(url, auth=auth, timeout=2)
+        logger.debug("[GET] Hit to %s", url)
+    except requests.exceptions.RequestException:
+        logger.warning("Failed to fetch data from %s", url)
+        return {}
+    return response.json()
+
+
+def validate_order(order: "Order") -> bool:
+    """Validate if checkout object contains enough information to generate a request to
+    avatax"""
+    if not order.lines.count():
+        return False
+    shipping_address = order.shipping_address
+    is_shipping_required = order.is_shipping_required()
+    address = shipping_address or order.billing_address
+
+    if not is_shipping_required and not address:
+        return False
+    if not shipping_address:
+        return False
+    if not order.shipping_method:
+        return False
+    return True
+
+
+def validate_checkout(checkout: "Checkout") -> bool:
```

Comment:
I wonder if we should put private the functions that are not part of the interface? It looked to me like that function was overriding the checkout workflow validators.